core/server/worker/src/main/java/alluxio/worker/netty/AbstractWriteHandler.java

```diff
@@ -121,21 +124,24 @@ public void channelRead(ChannelHandlerContext ctx, Object object) throws Excepti
     Protocol.WriteRequest writeRequest = msg.getMessage().asWriteRequest();
 
     try (LockResource lr = new LockResource(mLock)) {
-      boolean isPreviousContextNull = mContext == null;
-      if (isPreviousContextNull) {
-        // When mContext is null, create a new one as catching exceptions and replying errors
+      boolean isNewContextCreated = false;
+      if (mContext == null || mContext.isDoneUnsafe()) {
+        // We create a new context if the previous request completes (done flat is true) or the
```

Comment:
done flag?

Reply:
I wonder why not set `mContext` to null after the request completes (i.e. when `mContext.isDoneUnsafe` is true)? Then you can simplify this logic much

Reply:
what you are proposing is exactly the previous logic which introduces the race fixed in this PR. Consider this really rare case:

Server completes request A -> Server send OK response for request A -> Client receives OK response for request A -> Client sends new request B ->  Server receives request B  -> Server resets the context for request A.